src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -17,6 +16,256 @@ namespace Microsoft.Quantum.QsCompiler.Transformations.CallGraphWalker
+                {
+                    vInfo.LowLink = potentialMin;
+                    nodeInfo[node] = vInfo;
+                }
+            }
+
+            void strongconnect(int node)
+            {
+                // Set the depth index for node to the smallest unused index
+                nodeStack.Push(node);
+                nodeInfo[node] = (index, index, true);
+                index += 1;
+
+                // Consider children of node
+                foreach (var child in inputGraph[node])
+                {
+                    if (!nodeInfo.ContainsKey(child))
+                    {
+                        // Child has not yet been visited; recurse on it
+                        strongconnect(child);
+                        setMinLowLink(node, nodeInfo[child].LowLink);
+                    }
+                    else if (nodeInfo[child].OnStack)
+                    {
+                        // Child is in stack and hence in the current SCC
+                        // If child is not in stack, then (node, child) is an edge pointing to an SCC already found and must be ignored
+                        // Note: The next line may look odd - but is correct.
+                        // It says child.index not child.lowlink; that is deliberate and from the original paper
+                        setMinLowLink(node, nodeInfo[child].Index);
+                    }
+                }
+
+                // If node is a root node, pop the stack and generate an SCC
+                if (nodeInfo[node].LowLink == nodeInfo[node].Index)
+                {
+                    var scc = new HashSet<int>();
+
+                    var minNode = node;
+                    int nodeInScc;
+                    do
+                    {
+                        nodeInScc = nodeStack.Pop();
+                        var wInfo = nodeInfo[nodeInScc];
+                        wInfo.OnStack = false;
+                        nodeInfo[nodeInScc] = wInfo;
+                        scc.Add(nodeInScc);
+
+                        // Keep track of minimum node in scc
+                        if (minNode > nodeInScc)
+                        {
+                            minNode = nodeInScc;
+                        }
+
+                    } while (node != nodeInScc);
+                    SCCs.Add((scc, minNode));
+                }
+            }
+
+            foreach (var node in inputGraph.Keys)
+            {
+                if (!nodeInfo.ContainsKey(node))
+                {
+                    strongconnect(node);
+                }
+            }
+
+            return SCCs;
+        }
+
+        /// <summary>
+        /// Johnson's algorithm for finding all cycles in an SCC, or strongly-connected component.
+        ///
+        /// It will process an individual SCC for cycles by looking at each of their nodes
+        /// starting with the minimum-valued node of the SCC as the current node. Children of
```

Comment:
"minimum-valued" isn't very clear here. As far as I understand the SCC algorithm, it's the first node we discover in a component _given_ the specific traversal order of nodes in the input graph. Why starting from it is important? This SCC subgraph is a new dictionary and the traversal order of nodes in it doesn't necessarily match the traversal order in the overall input graph.

Reply:
It seemed to me, from the paper it is just a place to start. I'm not sure if there is any importance beyond that.

Reply:
That was my impression from the paper as well. Maybe, remove 'minimum-valued' from the comment to decrease confusion?